pkg/util/wait/backoff_test.go

```diff
@@ -0,0 +1,94 @@
+package wait
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+)
+
+type SpyTimer struct {
+	history *[]time.Duration
+	clock.Timer
+}
+
+func (s SpyTimer) Reset(d time.Duration) bool {
+	*s.history = append(*s.history, d)
+	return s.Timer.Reset(0)
+}
+
+func makeSpyTimer() SpyTimer {
```

Comment:
This is interesting, I didn't see this used before. Usually we use fakeClocks. I guess one advantage of a fake clock is that you don't need to wait at all, but given these times are short I think it is ok.

Reply:
But if we can use the fake clock instead of this mock function, I would prefer to use it.

Reply:
Note that I call Timer.Reset(0) on the real timer, so there is no waiting. I am open to using a fake clock, but I suppose I will still need to implement some functionality to capture the history (as I did with SpyTimer)?

Reply:
> Note that I call Timer.Reset(0) on the real timer, so there is no waiting.

Cool.

> I will still need to implement some functionality to capture the history (as I did with SpyTimer)?

Mostly in the tests I saw we are just using sleep to pass the arbitrary amount of time, like [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller_test.go#L4888-L4893).

Maybe fakeClock is more appropriate to test at the Scheduler level, where the internal details are more encapsulated. By analogy to the [Job controller](https://github.com/kubernetes/kubernetes/blob/9d87fa215d9e8020abdc17132d1252536cd752d2/pkg/controller/job/job_controller.go#L121) we could keep the clock at the Scheduler level, and use real for prod, but fake for testing.

Reply:
Yeah, not quite the same purpose as the fake clock.

But why do we even keep a timer inside the SpyTimer? It seems rather unused.

Reply:
> But why do we even keep a timer inside the SpyTimer? It seems rather unused.

To avoid implementing the rest of the clock.Timer interface